# Exploratory Data Analysis

In this EDA, I will be examining various data sources from [opendata.vancouver.ca](https://opendata.vancouver.ca/pages/home/).

## Imports

In [45]:
import numpy as np
import pandas as pd
import re
import requests

from matplotlib import pyplot as plt
from pyspark.sql import SparkSession

In [6]:
from pyspark.sql.functions import coalesce

In [7]:
from pyspark.ml.regression import LinearRegression
from pyspark.ml.feature import VectorAssembler
from pyspark.sql.functions import array, col, count, mean, sum, udf, when
from pyspark.sql.types import DoubleType, IntegerType, StringType, Row
from pyspark.sql.functions import sum, col, udf
from pyspark.sql.types import *

## Reading in the main dataset

In [8]:
spark = SparkSession.builder.appName('Ops').getOrCreate()

In [9]:
# allows for cleaner output
spark.conf.set("spark.sql.repl.eagerEval.enabled", True)

In [171]:
licences_df = (spark.read
          .option('header','true')
          .option('inferSchema','true')
          .option('sep', ';')
          .csv('data/business-licences.csv'))

In [172]:
licences_df.count()

443158

> There are 443 158 observations in this dataset. 

In [173]:
licences_df.printSchema()

root
 |-- FOLDERYEAR: integer (nullable = true)
 |-- LicenceRSN: integer (nullable = true)
 |-- LicenceNumber: string (nullable = true)
 |-- LicenceRevisionNumber: integer (nullable = true)
 |-- BusinessName: string (nullable = true)
 |-- BusinessTradeName: string (nullable = true)
 |-- Status: string (nullable = true)
 |-- IssuedDate: timestamp (nullable = true)
 |-- ExpiredDate: timestamp (nullable = true)
 |-- BusinessType: string (nullable = true)
 |-- BusinessSubType: string (nullable = true)
 |-- Unit: string (nullable = true)
 |-- UnitType: string (nullable = true)
 |-- House: string (nullable = true)
 |-- Street: string (nullable = true)
 |-- City: string (nullable = true)
 |-- Province: string (nullable = true)
 |-- Country: string (nullable = true)
 |-- PostalCode: string (nullable = true)
 |-- LocalArea: string (nullable = true)
 |-- NumberofEmployees: double (nullable = true)
 |-- FeePaid: integer (nullable = true)
 |-- ExtractDate: timestamp (nullable = true)
 |-- Geom: st

In [174]:
licences_df

+----------+----------+-------------+---------------------+--------------------+--------------------+--------------------+-------------------+-------------------+--------------------+---------------+----+--------+-----+-----------------+---------+--------+-------+----------+--------------------+-----------------+-------+-------------------+--------------------+
|FOLDERYEAR|LicenceRSN|LicenceNumber|LicenceRevisionNumber|        BusinessName|   BusinessTradeName|              Status|         IssuedDate|        ExpiredDate|        BusinessType|BusinessSubType|Unit|UnitType|House|           Street|     City|Province|Country|PostalCode|           LocalArea|NumberofEmployees|FeePaid|        ExtractDate|                Geom|
+----------+----------+-------------+---------------------+--------------------+--------------------+--------------------+-------------------+-------------------+--------------------+---------------+----+--------+-----+-----------------+---------+--------+-------+----------+--------------------+-----------------+-------+-------------------+--------------------+
|        15|   2380055|    15-150315|                    0|Brandi Nicole Eue...|Spa Haven Boutiqu...|Gone Out of Business|               null|               null|Therapeutic Touch...|           null|null|    null| null|             null|Vancouver|      BC|     CA|      null|Kensington-Cedar ...|              1.0|   null|2019-07-21 13:49:17|                null|
|        15|   2380056|    15-150316|                    0|Kelly Ashley Bake...|                null|             Pending|               null|               null|Therapeutic Touch...|           null|null|    null| null|             null|Vancouver|      BC|     CA|      null|           Kitsilano|              1.0|   null|2019-07-21 13:49:17|                null|
|        15|   2380058|    15-150318|                    0|Heart To Mind Cra...|       Heart To Mind|              Issued|2014-12-20 10:39:10|2015-12-31 00:00:00|Therapeutic Touch...|           null|null|    null| null|             null|Vancouver|      BC|     CA|      null|          Kerrisdale|              1.0|    133|2019-07-21 13:49:17|                null|
|        15|   2380062|    15-150322|                    0|         (Donna Sam)|                null|              Issued|2014-12-02 15:12:32|2015-12-31 00:00:00|Therapeutic Touch...|           null|null|    null| null|             null|Vancouver|      BC|     CA|      null|        South Cambie|              0.0|    155|2019-07-21 13:49:17|                null|
|        15|   2380065|    15-150325|                    0|James Joseph Thom...|  James Thompson RST|Gone Out of Business|               null|               null|Therapeutic Touch...|           null|null|    null| null|             null|Vancouver|      BC|     CA|      null|  Grandview-Woodland|              0.0|   null|2019-07-21 13:49:17|                null|
|        15|   2380066|    15-150326|                    0|   (Bunchu Praichit)|Bunchu Praichit T...|              Issued|2014-12-09 21:26:32|2015-12-31 00:00:00|Therapeutic Touch...|           null|null|    null| null|             null|Vancouver|      BC|     CA|      null|Kensington-Cedar ...|              0.0|    133|2019-07-21 13:49:17|                null|
|        15|   2380078|    15-150338|                    0| Tourland Travel Ltd|                null|              Issued|2014-11-24 10:30:17|2015-12-31 00:00:00|        Travel Agent|           null|   7|    Unit|  900|     W GEORGIA ST|Vancouver|      BC|     CA|   V6C 2W6|            Downtown|              2.0|    133|2019-07-21 13:49:17|"{""type"": ""Poi...|
|        15|   2380079|    15-150339|                    0|   Jetway Travel Inc|                null|              Issued|2015-02-27 14:16:59|2015-12-31 00:00:00|        Travel Agent|           null|null|    null| 1796|       RENFREW ST|Vancouver|      BC|     CA|   V5M 3H8|    Hastings-Sunrise|              2.0|    173|2019-07-21 13:49:17|"{""type"": ""Poi.

In [175]:
licences_df.groupBy('FOLDERYEAR').count()

+----------+-----+
|FOLDERYEAR|count|
+----------+-----+
|      null|    1|
|        13|60915|
|        16|61394|
|        20|62366|
|        94|    1|
|        96|   11|
|        19|70771|
|        15|60938|
|        14|60581|
|        18|66180|
+----------+-----+

In [176]:
licences_df.select('BusinessType').distinct().count()
licences_df.select('BusinessType').groupBy('BusinessType').count()

+--------------------+-----+
|        BusinessType|count|
+--------------------+-----+
|   Referral Services| 1835|
| Janitorial Services| 4366|
|Financial Institu...| 1480|
|Public Market Ope...|   20|
|  Non-profit Housing| 1674|
|               Motel|    7|
|   Printing Services| 1010|
|Retail Dealer - M...|   74|
|          Contractor|27394|
|    Private Hospital|   53|
|   Secondhand Dealer| 1279|
|       Talent Agency|  273|
|Entertainment Ser...| 2222|
|     Marine Services|  162|
|Sprinkler & Gas C...|    5|
|      Photo Services|  150|
|        Auto Repairs| 2082|
|       Tanning Salon|  113|
|      Gas Contractor| 1812|
|     Auto Wholesaler|  131|
+--------------------+-----+
only showing top 20 rows

## Initial Wrangling 

In [177]:
business_types = licences_df.select('BusinessType').distinct().collect()
business_list = [business_types[i].BusinessType for i in range(len(business_types))]

r = re.compile(".*[Ff]ood.*|.*[Rr]estaurant.*|.*[Ll]iquor.*")
newlist = list(filter(r.match, business_list)) 
print(newlist)

['Manufacturer - Food with Anc. Retail', 'Ltd Service Food Establishment', 'Temp Liquor Licence Amendment', 'Liquor Delivery Services', 'Liquor Establishment Standard', 'Retail Dealer - Food', 'Restaurant Class 2', 'Food Processing', 'Liquor Establishment Extended', 'Wholesale Dealer - Food with Anc. Retail', 'Restaurant Class 1', 'Liquor License Application', 'Warehouse Operator - Food', 'Manufacturer - Food', 'Liquor Retail Store', 'Wholesale Dealer - Food']


In [361]:
restaurants_1 = (licences_df.BusinessType == 'Restaurant Class 1')
restaurants_2 = (licences_df.BusinessType == 'Restaurant Class 2')
restaurants_3 = (licences_df.BusinessType == 'Ltd Service Food Establishment')


licences_rest_df = licences_df.filter(restaurants_1 | restaurants_2 | restaurants_3)

In [362]:
licences_rest_df.groupBy("Status").count()

+--------------------+-----+
|              Status|count|
+--------------------+-----+
|           Cancelled|  510|
|Gone Out of Business| 2488|
|              Issued|20666|
|            Inactive|  518|
|             Pending| 1380|
+--------------------+-----+

In [368]:
print(licences_rest_df.filter('BusinessTradeName is null').select('BusinessName').distinct().count())

924


> There are 924 businesses with missing name values so I will replace these entries with their business names. 

In [369]:
licences_rest_df = licences_rest_df.withColumn('BusinessTradeName', coalesce('BusinessTradeName', 'BusinessName'))

In [370]:
licences_rest_df = licences_rest_df.na.drop(subset = ['BusinessName'])

In [371]:
licences_rest_df

+----------+----------+-------------+---------------------+--------------------+--------------------+--------------------+-------------------+-------------------+--------------------+-------------------+----+--------+-----+--------------+---------+--------+-------+----------+--------------------+-----------------+-------+-------------------+--------------------+
|FOLDERYEAR|LicenceRSN|LicenceNumber|LicenceRevisionNumber|        BusinessName|   BusinessTradeName|              Status|         IssuedDate|        ExpiredDate|        BusinessType|    BusinessSubType|Unit|UnitType|House|        Street|     City|Province|Country|PostalCode|           LocalArea|NumberofEmployees|FeePaid|        ExtractDate|                Geom|
+----------+----------+-------------+---------------------+--------------------+--------------------+--------------------+-------------------+-------------------+--------------------+-------------------+----+--------+-----+--------------+---------+--------+-------+----------+--------------------+-----------------+-------+-------------------+--------------------+
|        14|   2382232|    14-306174|                    0|West Indian Tradi...|   Heirloom Juice Co|              Issued|2014-10-20 10:37:44|2014-12-31 00:00:00|Ltd Service Food ...|               null|null|    null| 2861|  GRANVILLE ST|Vancouver|      BC|     CA|   V6H 3J4|            Fairview|              0.0|    170|2019-07-21 13:49:17|"{""type"": ""Poi...|
|        14|   2382785|    14-306719|                    0|      1012703 BC Ltd|My Toan Vietnames...|              Issued|2014-10-29 08:54:12|2014-12-31 00:00:00|Ltd Service Food ...|               null| F1A|    Unit|  650|     W 41ST AV|Vancouver|      BC|     CA|   V5Z 2M9|            Oakridge|              7.0|    170|2019-07-21 13:49:17|"{""type"": ""Poi...|
|        14|   2382833|    14-306765|                    0|What's Up? Hot Do...|What's Up? Hot Do...|             Pending|               null|               null|  Restaurant Class 1|  No Liquor Service|null|    null| 2481| E HASTINGS ST|Vancouver|      BC|     CA|   V5K 1Y8|    Hastings-Sunrise|              2.0|     51|2019-07-21 13:49:17|"{""type"": ""Poi...|
|        14|   2383108|    14-307037|                    0|      0949813 BC Ltd|             Sushi 7|              Issued|2014-11-05 11:34:48|2014-12-31 00:00:00|  Restaurant Class 1|With Liquor Service|null|    null| 3488|     CAMBIE ST|Vancouver|      BC|     CA|   V5Z 2W8|          Riley Park|              0.0|    130|2019-07-21 13:49:17|"{""type"": ""Poi...|
|        14|   2384487|    14-146246|                    1|Don Maker Enterpr...|     Oak & 16th Cafe|              Issued|2014-10-27 13:29:53|2014-12-31 00:00:00|  Restaurant Class 1|  No Liquor Service|null|    null| 3208|        OAK ST|Vancouver|      BC|     CA|   V6H 2L3|        South Cambie|              2.0|     10|2019-07-21 13:49:17|"{""type"": ""Poi...|
|        14|   2385125|    14-309023|                    0|The Army Navy & A...|        ANAF Unit 26|              Issued|2014-11-26 10:50:19|2014-12-31 00:00:00|  Restaurant Class 2|With Liquor Service|null|    null|  727|     E 49TH AV|Vancouver|      BC|     CA|   V5W 2H3|              Sunset|              0.0|    567|2019-07-21 13:49:17|"{""type"": ""Poi...|
|        14|   2385202|    14-309099|                    0|Soffee Coffee &Te...|   Soffee Tea & Cafe|             Pending|               null|               null|Ltd Service Food ...|               null|null|    null| 1426|      W 8TH AV|Vancouver|      BC|     CA|   V6H 1E1|            Fairview|              4.0|    130|2019-07-21 13:49:17|"{""type"": ""Poi...|
|        14|   2385663|    14-309548|                    0|      0992799 BC Ltd|           One Under|              Issued|2014-10-30 15:30:56|2014-12-31 00:00:00|  Restaurant Class 1|With Liquor Service|null|    null|  470|  GRANVILLE ST|Vancouver|      BC|     CA|   V6C 1V5|            Downtown|             20.0|    345|2019-07-21 13:49:17|"{""typ

In [372]:
licences_rest_df.count()

25538

In [313]:
from pyspark.sql.functions import regexp_extract, col

# Search for names using regex

search_term = 'Spot'
licences_rest_df.filter(licences_rest_df.BusinessTradeName.rlike('.*{}.*'.format(search_term)))


+----------+----------+-------------+---------------------+--------------------+--------------------+------+-------------------+-------------------+--------------------+-------------------+----+--------+-----+---------------+---------+--------+-------+----------+--------------------+-----------------+-------+-------------------+--------------------+
|FOLDERYEAR|LicenceRSN|LicenceNumber|LicenceRevisionNumber|        BusinessName|   BusinessTradeName|Status|         IssuedDate|        ExpiredDate|        BusinessType|    BusinessSubType|Unit|UnitType|House|         Street|     City|Province|Country|PostalCode|           LocalArea|NumberofEmployees|FeePaid|        ExtractDate|                Geom|
+----------+----------+-------------+---------------------+--------------------+--------------------+------+-------------------+-------------------+--------------------+-------------------+----+--------+-----+---------------+---------+--------+-------+----------+--------------------+-----------------+-------+-------------------+--------------------+
|        16|   2603928|    16-145091|                    0|  White Spot Limited|White Spot Restau...|Issued|2015-12-24 11:57:18|2016-12-31 00:00:00|  Restaurant Class 1|With Liquor Service|null|    null|  650|      W 41st Av|Vancouver|      BC|     CA|    V5Z2M9|            Oakridge|              0.0|   2015|2019-07-21 13:49:20|"{""type"": ""Poi...|
|        16|   2604217|    16-145379|                    0|Top Edge Holdings...|White Spot Restau...|Issued|2015-12-10 13:21:40|2016-12-31 00:00:00|  Restaurant Class 1|With Liquor Service|null|    null| 5367| WEST BOULEVARD|Vancouver|      BC|     CA|   V6M 3W4|       Arbutus-Ridge|             40.0|   1806|2019-07-21 13:49:20|"{""type"": ""Poi...|
|        16|   2604226|    16-145388|                    0|Pacific Internati...|White Spot Restau...|Issued|2015-11-12 12:33:43|2016-12-31 00:00:00|  Restaurant Class 1|  No Liquor Service|null|    null| 1881|    Cornwall Av|Vancouver|      BC|     CA|   V6J 1C6|           Kitsilano|             10.0|    714|2019-07-21 13:49:20|"{""type"": ""Poi...|
|        16|   2604241|    16-145403|                    0|No 109 Seabright ...|White Spot Restau...|Issued|2015-11-05 13:24:16|2016-12-31 00:00:00|  Restaurant Class 1|With Liquor Service|null|    null| 1476|       Kingsway|Vancouver|      BC|     CA|   V5N 2R5|Kensington-Cedar ...|             49.0|   1688|2019-07-21 13:49:20|"{""type"": ""Poi...|
|        16|   2604304|    16-145466|                    0|White Spot Limite...|White Spot Restau...|Issued|2015-12-24 11:59:33|2016-12-31 00:00:00|  Restaurant Class 1|With Liquor Service|null|    null| 1126|   SE Marine Dr|Vancouver|      BC|     CA|   V5X 2V7|              Sunset|              0.0|   1588|2019-07-21 13:49:20|                null|
|        16|   2604901|    16-146062|                    0|Mayfair Propertie...|          White Spot|Issued|2016-01-06 11:29:37|2016-12-31 00:00:00|  Restaurant Class 1|With Liquor Service|null|    null|  712|       DRAKE ST|Vancouver|      BC|     CA|   V6Z 2W6|            Downtown|              0.0|   2443|2019-07-21 13:49:20|"{""type"": ""Poi...|
|        16|   2605056|    16-146217|                    0|Donair Spot Resta...|         Donair Spot|Issued|2015-12-29 15:13:48|2016-12-31 00:00:00|  Restaurant Class 1|  No Liquor Service|null|    null|  806|     W BROADWAY|Vancouver|      BC|     CA|   V5Z 1J8|            Fairview|              1.0|    714|2019-07-21 13:49:20|"{""type"": ""Poi...|
|        16|   2605341|    16-146502|                    0|  White Spot Limited|White Spot Restau...|Issued|2015-12-24 12:00:10|2016-12-31 00:00:00|  Restaurant Class 1|With Liquor Service|null|    null|  405|    DUNSMUIR ST|Vancouver|      BC|     CA|   V6B 1X4|            Downtown|            100.0|   3326|2019-07-21 13:49:20|"{""type"": ""Poi...|
|        16|   2584230|    16-125412|                    0|Donair Spot Resta...|         Donair Spot|Issued|2015-12-29 15:17:02|

In [296]:
test[0].Geom

'"{""type"": ""Point"", ""coordinates"": [-123.156080648747, 49.2321417650711]}"'

> How many stores have changed owners? (BusinessName is the owner, BusinessTradeName is the name of the shop). There are 5682 different owners and 5150 different restaurants. 

In [203]:
# THis is definitely wrong
print(licences_rest_df.select('BusinessName').distinct().count())
print(licences_rest_df.select('BusinessTradeName').distinct().count())

5681
5148


> Use an external dataset such as yelp or zomato to get restuarant cuisine, price range, and other features. 

In [403]:
# use this table to populate the features from yelp

rest_distinct = licences_rest_df.select(['BusinessTradeName', 'City', 'Geom', 'PostalCode', 'House', 'Street']).distinct()

In [404]:
rest_distinct.count()

6353

> There are 6353 distinct restaurants locations.

In [410]:
print(rest_distinct.filter('Geom is null').filter('PostalCode is null').filter('House is null').filter('Street is null').count())

4


> 4 of these restaurants do not have any information on the location. Since this is a very small proportion of the data, I will remove them.

## YELP API

https://www.yelp.ca/developers/documentation/v3/get_started   
https://github.com/Yelp/yelp-fusion#code-samples


In [256]:
api_key = '2MHpkRrNYZCK-R3gsH-cWD4XxA9iBgZsUCpa_meN4QK2KcstC0BPrCcAGCyCcvkrZ_UxbFX65IwwKibVR6earNojMaB6FlAkKFnmJqh72_GKqgyjk0JCGTmtBl6KXnYx'

In [263]:
headers = {
    'Authorization': 'Bearer %s' % api_key,
}

In [264]:
headers

{'Authorization': 'Bearer 2MHpkRrNYZCK-R3gsH-cWD4XxA9iBgZsUCpa_meN4QK2KcstC0BPrCcAGCyCcvkrZ_UxbFX65IwwKibVR6earNojMaB6FlAkKFnmJqh72_GKqgyjk0JCGTmtBl6KXnYx'}

In [348]:
params = {'apikey': '2MHpkRrNYZCK-R3gsH-cWD4XxA9iBgZsUCpa_meN4QK2KcstC0BPrCcAGCyCcvkrZ_UxbFX65IwwKibVR6earNojMaB6FlAkKFnmJqh72_GKqgyjk0JCGTmtBl6KXnYx',
          'term': 'Fiery Pizza',
        #  'location': '1126 SE Marine Dr V5X 2V7',
        #  'longitude': -123.156010,
        #  'latitude': 49.237763,
          'sort_by': 'distance'}
requests.get('https://api.yelp.com/v3/businesses/search', headers = headers, params = params).json()

{'error': {'code': 'VALIDATION_ERROR',
  'description': 'Please specify a location or a latitude and longitude'}}

## ZOMATO API

https://developers.zomato.com/api  
https://github.com/RapidSoftwareSolutions/Marketplace-Zomato-Package   
api key: 99beac6a17fef3d9816ec61eb532bab0

In [169]:
import os

print(os.getenv('ZOMATO_API_KEY'))

None


In [141]:
# categories
params = {'apikey': '99beac6a17fef3d9816ec61eb532bab0',
          'city_id': None}
categories = requests.get('https://developers.zomato.com/api/v2.1/categories', params = params).json()


In [143]:
# cuisines
params = {'apikey': '99beac6a17fef3d9816ec61eb532bab0',
          'city_id': 256}
cuisines = requests.get('https://developers.zomato.com/api/v2.1/cuisines', params = params).json()


In [145]:
# types
params = {'apikey': '99beac6a17fef3d9816ec61eb532bab0',
          'city_id': 256}
types = requests.get('https://developers.zomato.com/api/v2.1/establishments', params = params).json()

In [166]:
# restaurants
params = {'apikey': '99beac6a17fef3d9816ec61eb532bab0',
          'entity_id': 256,
          'entity_type': 'city',
          'start': 80,
          'count': 100}
restaurants = requests.get('https://developers.zomato.com/api/v2.1/search', params = params).json()

In [170]:
requests.get('https://api.yelp.com/v3/businesses/search')

<Response [400]>